[Back to questions](../README.md)

## Solution to [74d2](../questions/74d2): *Exceptions and inheritance (i)*

See code at `solutions/code/tutorialquestions/question74d2`

\begin{question}{Exceptions and inheritance (i).}\label{question:inheritance}
%
Classes \lstinline{A} and \lstinline{B} \emph{do} compile.  The potential reason for which I was hoping you might have
thought they would not compile was the fact that \lstinline{A} declares a method \lstinline{foo} with the following signature:

\begin{lstlisting}
public void foo() throws Exception;
\end{lstlisting}
%
while \lstinline{B} attempts to override \lstinline{foo} with a different signature:

\begin{lstlisting}
@Override
public void foo();
\end{lstlisting}

The difference is that \lstinline{foo} in \lstinline{A} throws \lstinline{Exception} while \lstinline{foo} in \lstinline{B} does not throw any exception.  This is OK, however: it is permissible for an overridden method to throw a \emph{narrower} type of exception than the exception specified in the superclass, or
(as in this case) to throw no exception at all.

The way to think about this is that it is OK for the overridden method to behave \emph{better} than the superclass method.  Client code that uses the superclass must be prepared to deal with any exceptions the superclass method may throw.  If at runtime an instance of the subclass, whose method behaves better, is present, this is no problem.  (Do you see why the rule therefore could not be the other way around?  I.e., why wouldn't it make sense to allow subclass methods to throw \emph{more}, or \emph{more general} exceptions than the superclass specifies?)

Changing to body of \lstinline{foo} in \lstinline{B} from:

\begin{lstlisting}
@Override
public void foo() {

}
\end{lstlisting}
%
to:
\begin{lstlisting}
@Override
public void foo() {
  super.foo();
}
\end{lstlisting}
%
leads to a compilation error.  The reason is that \lstinline{super.foo();} requests that method \lstinline{foo} in \lstinline{A} is called.  However, this method may throw an \lstinline{Exception}, which \lstinline{foo} in \lstinline{B} would have to deal with, but does not.

This problem can be solved either by changing \lstinline{foo} in \lstinline{B} to throw \lstinline{Exception}, so that if \lstinline{foo} in \lstinline{A} does throw an \lstinline{Exception} this is \emph{propagated} by \lstinline{foo} in \lstinline{B}, or by enclosing the call to \lstinline{super.foo();} in a \lstinline{try...catch} block:

\begin{lstlisting}
@Override
public void foo() {
  try {
    super.foo();
  } catch(Exception e) {
    // Do something
  }
}
\end{lstlisting}

\end{question}
